### PR TITLE
feat(sdk-coin-avaxp): implement tx builder for import in c chain

### PR DESCRIPTION
**Pull request by jrafael-simtlix**
Features:
- C chain fee is a Ethereum model with  fee rate (price), size (cost or units) and fee (price * units). 
- Transaction could be EVMTx or PVMTx with its respective base tx
- add test folder in tsconfig
- TransferableInput and Output are other clases

The last items force to decouple createInputsOutputs method.
- the createInputOutput is replaced with a utxoToInput mapper. 
- the ImportInCTxBuilder take that input and instance a TransferableInput of C chain.
- the signature is improved to handler N of M signatures
- RecoveryMode backwards compatibility is guaranteed. 

[BG-55033](https://bitgoinc.atlassian.net/browse/BG-55033)

### CODE REVIEW

**Review by lcovar — DISMISSED** [review]
lgtm

File: modules/sdk-coin-avaxp/src/lib/utxoEngine.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
+import { DecodedUtxoObj, SECP256K1_Transfer_Output } from './iface';
```
Comment (joey-bai) [inline modules/sdk-coin-avaxp/src/lib/utxoEngine.ts:1]: do we need unit tests for this file?

File: modules/sdk-coin-avaxp/src/lib/utils.ts [hunk 2]
```diff
@@ -274,18 +269,29 @@ export class Utils implements BaseUtils {
...
-        .map((a) => this.addressToString(network.hrp, network.alias, a))
-        .sort()
-        .join('~');
-      return {
-        value: amountOutput.getAmount().toString(),
-        address,
-      };
+  mapOutputToEntry(network: AvalancheNetwork): (Output) => Entry {
+    return (output: Output) => {
+      if ('getOutput' in output) {
```
Comment (joey-bai) [inline modules/sdk-coin-avaxp/src/lib/utils.ts:274]: this is hacky. can we have a better typing for this? 

maybe something like:
```
// https://www.typescriptlang.org/docs/handbook/advanced-types.html
private static isTransferableOutput(output: Output | TransferableOutput): output is TransferableOutput {
    return (output as TransferableOutput).getOutput !== undefined;

mapOutputToEntry(network: AvalancheNetwork): (Output | TransferableOutput) => Entry {
    return (output: Output | TransferableOutput) => {
        if (isTransferableOutput) {
            ...
        } else {
            ...
```

File: modules/sdk-coin-avaxp/src/lib/utils.ts [hunk 3]
```diff
@@ -274,18 +269,29 @@ export class Utils implements BaseUtils {
...
+          .join('~');
+        return {
+          value: amountOutput.getAmount().toString(),
+          address,
+        };
+      } else {
+        const evmOuput = output as EVMOutput;
+        return {
+          // it should be evmOuput.getAmount()
+          value: new BN((evmOuput as any).amount).toString(),
```
Comment (joey-bai) [inline modules/sdk-coin-avaxp/src/lib/utils.ts:290]: not sure what this comment means. but why can't we use `evmOutput.getAmount()`?

Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/utils.ts:290]: `evmOutput.getAmount()` return 0. 
I think that avalanchejs should fix it.

Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/utils.ts:290]: https://github.com/ava-labs/avalanchejs/issues/652

File: modules/sdk-coin-avaxp/src/lib/utxoEngine.ts (hunk 1 above)
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/utxoEngine.ts:1]: done

File: modules/sdk-coin-avaxp/src/lib/utils.ts (hunk 2 above)
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/utils.ts:274]: done

File: modules/sdk-coin-avaxp/src/lib/transaction.ts [hunk 4]
```diff
@@ -168,14 +165,15 @@ export class Transaction extends BaseTransaction {
     if (!this.avaxPTransaction) {
       throw new InvalidTransactionError('Empty transaction data');
     }
+    const memo = 'getMemo' in this.avaxPTransaction ? utils.bufferToString(this.avaxPTransaction.getMemo()) : undefined;
```
Comment (CapnMigraine) [inline modules/sdk-coin-avaxp/src/lib/transaction.ts:168]: in which case is `getMemo()` not defined?
please add a comment to the code explaining

File: modules/sdk-coin-avaxp/src/lib/transaction.ts [hunk 5]
```diff
@@ -238,23 +236,30 @@ export class Transaction extends BaseTransaction {
     }
   }
 
+  /**
+   * Get a TransasctionFee.
```
Comment (CapnMigraine) [inline modules/sdk-coin-avaxp/src/lib/transaction.ts:240]: nit: typo
```suggestion
   * Get a Transaction Fee.
```

File: modules/sdk-coin-avaxp/src/lib/transactionBuilderFactory.ts [hunk 6]
```diff
@@ -17,19 +19,46 @@ export class TransactionBuilderFactory extends BaseTransactionBuilderFactory {
...
+      tx = new Tx();
+      // could throw an error if a txType doesn't match.
+      tx.fromBuffer(BufferAvax.from(raw, 'hex'));
+
+      const sameBlockchainId =
+        utils.cb58Encode(tx.getUnsignedTx().getTransaction().getBlockchainID()) ===
+        (this._coinConfig.network as AvalancheNetwork).blockchainID;
+      if (!sameBlockchainId) {
+        throw new Error('It is not a transaction of this network');
+      }
```
Comment (CapnMigraine) [inline modules/sdk-coin-avaxp/src/lib/transactionBuilderFactory.ts:34]: Duplicated code down in ln:49. This validation could be done inside a different try/catch block do avoid duplication.

File: modules/sdk-coin-avaxp/src/lib/transactionBuilderFactory.ts [hunk 7]
```diff
@@ -17,19 +19,46 @@ export class TransactionBuilderFactory extends BaseTransactionBuilderFactory {
...
+
+      if (ValidatorTxBuilder.verifyTxType(tx.getUnsignedTx().getTransaction())) {
+        transactionBuilder = this.getValidatorBuilder();
+      } else if (ExportTxBuilder.verifyTxType(tx.getUnsignedTx().getTransaction())) {
+        transactionBuilder = this.getExportBuilder();
+      } else if (ImportTxBuilder.verifyTxType(tx.getUnsignedTx().getTransaction())) {
+        transactionBuilder = this.getImportBuilder();
+      } else {
+        throw new NotSupported('Transaction cannot be parsed or has an unsupported transaction type');
+      }
```
Comment (CapnMigraine) [inline modules/sdk-coin-avaxp/src/lib/transactionBuilderFactory.ts:44]: Exception caught locally, duplicated code

File: modules/sdk-coin-avaxp/src/lib/utils.ts [hunk 8]
```diff
@@ -269,23 +271,37 @@ export class Utils implements BaseUtils {
     }
   }
 
+  isTransferableOutput(output: Output): output is TransferableOutput {
```
Comment (CapnMigraine) [inline modules/sdk-coin-avaxp/src/lib/utils.ts:274]: Add jsdoc

File: modules/sdk-coin-avaxp/src/lib/utils.ts [hunk 9]
```diff
@@ -269,23 +271,37 @@ export class Utils implements BaseUtils {
...
+          .getAddresses()
+          .map((a) => this.addressToString(network.hrp, network.alias, a))
+          .sort()
+          .join('~');
+        return {
+          value: amountOutput.getAmount().toString(),
+          address,
+        };
+      } else {
+        const evmOuput = output as EVMOutput;
```
Comment (CapnMigraine) [inline modules/sdk-coin-avaxp/src/lib/utils.ts:297]: typo
```suggestion
        const evmOutput = output as EVMOutput;
```

File: modules/sdk-coin-avaxp/src/lib/utxoEngine.ts [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+import { BN, Buffer as BufferAvax } from 'avalanche';
+import { Signature } from 'avalanche/dist/common';
+import utils from './utils';
+import { BuildTransactionError } from '@bitgo/sdk-core';
+import { StandardAmountInput, StandardTransferableInput } from 'avalanche/dist/common/input';
+
+export interface InputData {
+  amount: BN;
+  txidBuf: BufferAvax;
+  outputidx: BufferAvax;
```
Comment (CapnMigraine) [inline modules/sdk-coin-avaxp/src/lib/utxoEngine.ts:11]: consistency-nit: use camelCase
```suggestion
  outputIdx: BufferAvax;
```

File: modules/sdk-coin-avaxp/src/lib/utxoEngine.ts [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+  signareIdx: number[];
+  signatures: Signature[];
+}
+
+/**
+ * Inputs can be controlled but outputs get reordered in transactions
+ * In order to make sure that the mapping is always correct we create an addressIndx which matches to the appropiate
+ * signatureIdx
+ * @param {TransferableInput[]} utxos as transaction ins.
```
Comment (CapnMigraine) [inline modules/sdk-coin-avaxp/src/lib/utxoEngine.ts:20]: param type doesn't match the function signature

File: modules/sdk-coin-avaxp/src/lib/utxoEngine.ts [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+  amount: BN;
+  txidBuf: BufferAvax;
+  outputidx: BufferAvax;
+  signareIdx: number[];
+  signatures: Signature[];
+}
+
+/**
+ * Inputs can be controlled but outputs get reordered in transactions
+ * In order to make sure that the mapping is always correct we create an addressIndx which matches to the appropiate
```
Comment (CapnMigraine) [inline modules/sdk-coin-avaxp/src/lib/utxoEngine.ts:18]: nit: typo
```suggestion
 * In order to make sure that the mapping is always correct we create an addressIndex which matches to the appropriate
```

File: modules/sdk-coin-avaxp/src/lib/utxoEngine.ts [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+      addresses: [], // this is empty since the inputs from deserialized transaction don't contain addresses
+      addressesIndex,
+    };
+  });
+}
+
+/**
+ * Convert Utxos into inputs data. Input Objects changes regarding chains. This method return a plain object to be mapped late in chain input.
+ * Sender is a list of owners address and utxo address must contains all of them.
+ * Signers is a list of sender cat it in threshold size. Firsts senders are the signers.
```
Comment (CapnMigraine) [inline modules/sdk-coin-avaxp/src/lib/utxoEngine.ts:45]: typo on `sender cat it`?

File: modules/sdk-coin-avaxp/src/lib/utxoEngine.ts [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+): {
+  inputs: InputData[];
+  amount: BN;
+} {
+  // amount spent so far
+  let currentTotal: BN = new BN(0);
+
+  const inputs = utxos
+    .filter((utxo) => utxo && utxo.outputID === SECP256K1_Transfer_Output)
+    .map((utxo, i) => {
```
Comment (CapnMigraine) [inline modules/sdk-coin-avaxp/src/lib/utxoEngine.ts:75]: `i` is not used

File: modules/sdk-coin-avaxp/src/lib/importInCTxBuilder.ts [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+      throw new Error('AssetID are not equals');
+    }
+    this.transaction._to = [output.getAddress()];
+
+    const input = baseTx.getImportInputs();
+
+    this.transaction._utxos = recoverUtxos(input);
+
+    const totalInputAmount = input.reduce((t, i) => t.add((i.getInput() as AmountInput).getAmount()), new BN(0));
+    // it should be (output as AmountOutput).getAmount(), but it's not working.
```
[the remark this answers is not in the document]
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/importInCTxBuilder.ts:82]: It's a issue in Avalanche js.
https://github.com/ava-labs/avalanchejs/issues/652

File: modules/sdk-coin-avaxp/src/lib/importInCTxBuilder.ts [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+import utils from './utils';
+import { BN, Buffer as BufferAvax } from 'avalanche';
+import { Credential } from 'avalanche/dist/common';
+import { Transaction } from './transaction';
+import { AtomicTransactionBuilder } from './atomicTransactionBuilder';
+import { recoverUtxos, utxoToInput } from './utxoEngine';
+import { BaseTx, Tx } from './iface';
+
+export class ImportInCTxBuilder extends AtomicTransactionBuilder {
+  private static _targetChainId: BufferAvax = utils.cb58Decode('yH8D7ThNJkxmtkuv2jgBa4P1Rn3Qpr4pPr7QYNfcdoS6k6HWp');
```
[the remark this answers is not in the document]
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/importInCTxBuilder.ts:24]: Yes, it does. But I forgotten to remove this unnecessary field

File: modules/sdk-coin-avaxp/src/lib/utxoEngine.ts (hunk 14 above)
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/utxoEngine.ts]: done 
https://github.com/BitGo/BitGoJS/blob/6d023175be2d689b599bbc8e62e71c713fa7934a/modules/sdk-coin-avaxp/src/lib/utxoEngine.ts#L75

File: modules/sdk-coin-avaxp/src/lib/utxoEngine.ts (hunk 13 above)
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/utxoEngine.ts]: https://github.com/BitGo/BitGoJS/blob/6d023175be2d689b599bbc8e62e71c713fa7934a/modules/sdk-coin-avaxp/src/lib/utxoEngine.ts#L45

File: modules/sdk-coin-avaxp/src/lib/utxoEngine.ts (hunk 11 above)
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/utxoEngine.ts]: https://github.com/BitGo/BitGoJS/blob/6d023175be2d689b599bbc8e62e71c713fa7934a/modules/sdk-coin-avaxp/src/lib/utxoEngine.ts#L20

File: modules/sdk-coin-avaxp/src/lib/utxoEngine.ts (hunk 10 above)
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/utxoEngine.ts]: https://github.com/BitGo/BitGoJS/blob/6d023175be2d689b599bbc8e62e71c713fa7934a/modules/sdk-coin-avaxp/src/lib/utxoEngine.ts#L11

File: modules/sdk-coin-avaxp/src/lib/importInCTxBuilder.ts [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+  /** @inheritdoc */
+  protected fromImplementation(rawTransaction: string): Transaction {
+    const tx = new EVMTx();
+    tx.fromBuffer(BufferAvax.from(rawTransaction, 'hex'));
+    this.initBuilder(tx);
+    return this.transaction;
+  }
+
+  /**
+   * Create inputs by mapping {@see utxoEngiene.utxoToInput} result.
```
[the remark this answers is not in the document]
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/importInCTxBuilder.ts]: https://github.com/BitGo/BitGoJS/blob/6d023175be2d689b599bbc8e62e71c713fa7934a/modules/sdk-coin-avaxp/src/lib/importInCTxBuilder.ts#L154

File: modules/sdk-coin-avaxp/src/lib/importInCTxBuilder.ts [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+    }
+    this.transaction._to = [output.getAddress()];
+
+    const input = baseTx.getImportInputs();
+
+    this.transaction._utxos = recoverUtxos(input);
+
+    const totalInputAmount = input.reduce((t, i) => t.add((i.getInput() as AmountInput).getAmount()), new BN(0));
+    // it should be (output as AmountOutput).getAmount(), but it's not working.
+    const totalOutpuAmount = new BN((output as any).amount);
```
[the remark this answers is not in the document]
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/importInCTxBuilder.ts]: https://github.com/BitGo/BitGoJS/blob/6d023175be2d689b599bbc8e62e71c713fa7934a/modules/sdk-coin-avaxp/src/lib/importInCTxBuilder.ts#L81

File: modules/sdk-coin-avaxp/src/lib/importInCTxBuilder.ts [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+    ) {
+      throw new Error('Network or blockchain is not equals');
+    }
+
+    if (!this.verifyTxType(baseTx)) {
+      throw new NotSupported('Transaction cannot be parsed or has an unsupported transaction type');
+    }
+
+    // The outputs is a signler C-Chain address result.
+    // It's expected to have only one outputs.
```
[the remark this answers is not in the document]
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/importInCTxBuilder.ts]: https://github.com/BitGo/BitGoJS/blob/6d023175be2d689b599bbc8e62e71c713fa7934a/modules/sdk-coin-avaxp/src/lib/importInCTxBuilder.ts#L63

File: modules/sdk-coin-avaxp/src/lib/transactionBuilderFactory.ts [hunk 20]
```diff
@@ -17,19 +19,46 @@ export class TransactionBuilderFactory extends BaseTransactionBuilderFactory {
...
+    } catch {
+      tx = new EVMTx();
+      tx.fromBuffer(BufferAvax.from(raw, 'hex'));
 
+      const sameBlockchainId =
+        utils.cb58Encode(tx.getUnsignedTx().getTransaction().getBlockchainID()) ===
+        (this._coinConfig.network as AvalancheNetwork).cChainBlockchainID;
+      if (!sameBlockchainId) {
+        throw new Error('It is not a transaction of this network or C chain');
+      }
```
Comment (lcovar) [inline modules/sdk-coin-avaxp/src/lib/transactionBuilderFactory.ts:54]:
```suggestion
validateBlockChainId(tx)
```

Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/transactionBuilderFactory.ts:54]: it'll be done in next pr
https://github.com/BitGo/BitGoJS/blob/b3805da4d5e69b358be6ad5e813e5e57b83b3c0e/modules/sdk-coin-avaxp/src/lib/transactionBuilderFactory.ts#L30

File: modules/sdk-coin-avaxp/src/lib/transactionBuilderFactory.ts (hunk 6 above)
Comment (CapnMigraine) [reply modules/sdk-coin-avaxp/src/lib/transactionBuilderFactory.ts:34]: Please re-write this to remove code duplication. You can split it into 2 try/catch blocks.

File: modules/sdk-coin-avaxp/src/lib/transaction.ts (hunk 4 above)
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/transaction.ts:169]: done

File: modules/sdk-coin-avaxp/src/lib/utils.ts (hunk 8 above)
Comment (jrafael-simtlix) [reply modules/sdk-coin-avaxp/src/lib/utils.ts:272]: done

**Review by CapnMigraine — APPROVED** [review]

**Review by byhow — APPROVED** [review]

**Pull request merged**